### PR TITLE
[CWS] Autolearn anomalies

### PR DESCRIPTION
**Pull request by Gui774ume**
### What does this PR do?

This PR has 2 main goals:
- it introduces the `ActivityTree` structure, a generic structure used both by Activity Dumps and Security Profiles to share the same tree logic instead of duplicating it. This shared structure has
- It also introduces the ability to inject anomalies into Security Profiles. This mechanism is used with a redemption period used to prevent our Security Profile from triggering anomalies for newly created workloads.

4 new configuration parameters were introduced:
- `runtime_security_config.security_profile.anomaly_detection.event_types` : defines the list of events that should be allowed to generate anomaly detections
- `runtime_security_config.security_profile.anomaly_detection.minimum_stable_period` : defines the minimum amount of time during which the events that diverge from their profiles are automatically added in their profiles without triggering an anomaly detection event
- `runtime_security_config.security_profile.anomaly_detection.unstable_profile_time_threshold` : defines the maximum amount of time to wait until a profile that hasn't reached a stable state is considered as unstable
- `runtime_security_config.security_profile.anomaly_detection.unstable_profile_size_threshold` : defines the maximum size a profile can reach past which it is considered unstable

### Motivation

This is an iterative step for the Security Profile and Anomaly detection feature for CWS. CWS is now able to generate anomalies based on unusual workload behavior, instead of simply relying on hardcoded rules.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by brycekahle — APPROVED** [review]
shallow review for `pkg/config/system_probe.go`

File: pkg/security/probe/probe_linux.go [hunk 1]
```diff
@@ -321,16 +321,17 @@ func (p *Probe) handleAnomalyDetection(event *model.Event) {
 		return
 	}
 
-	if !profile.IsAnomalyDetectionEvent(event.GetEventType()) {
+	if profile.IsAnomalyDetectionEvent(event.GetEventType()) {
```
Comment (safchain) [inline pkg/security/probe/probe_linux.go:324]: why this `if` and the `else` ? why not using as previously the `IsAnomalyDetectionEvent` event ?

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+// nolint: unused
+func (at *ActivityTree) debug(w io.Writer) {
+	for _, root := range at.ProcessNodes {
+		root.debug(w, "")
+	}
+}
+
+// ScrubAndRetainProcessArgsEnvs scrubs and retains process args and envs
+func (at *ActivityTree) ScrubAndRetainProcessArgsEnvs(resolver *process.Resolver) {
```
Comment (safchain) [inline pkg/security/security_profile/activity_tree/activity_tree.go:122]: now should be `ScrubProcessArgsEnvs` . we don't need to retain anymore

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+		return node.InsertSyscalls(event, at.SyscallsMask)
+	}
+
+	return false, nil
+}
+
+// CreateProcessNode finds or a create a new process activity node in the activity dump if the entry
+// matches the activity dump selector.
+func (at *ActivityTree) CreateProcessNode(entry *model.ProcessCacheEntry, generationType NodeGenerationType, shadowInsertion bool) (node *ProcessNode, newProcessNode bool, err error) {
+	if entry == nil {
```
Comment (safchain) [inline pkg/security/security_profile/activity_tree/activity_tree.go:259]: I'm not sure this line is required, we should always have here a valid entry

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+		return nil, false, nil
+	}
+
+	// drop processes with abnormal paths
+	if entry.GetPathResolutionError() != "" {
```
Comment (safchain) [inline pkg/security/security_profile/activity_tree/activity_tree.go:264]: same here see the line : https://github.com/DataDog/datadog-agent/pull/16744/files#diff-c7e94094f2ca1d0e250b89fa8f9cd016c90791b1c9ca72b7b376c27bc6ea4492R161

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+		// if a node was found, and if the entry has a valid cookie, insert a cookie shortcut
+		if entry.Cookie > 0 && node != nil {
+			at.CookieToProcessNode[entry.Cookie] = node
+		}
+	}()
+
+	// find or create a ProcessActivityNode for the parent of the input ProcessCacheEntry. If the parent is a fork entry,
+	// jump immediately to the next ancestor.
+	parentNode, newProcessNode, err := at.CreateProcessNode(entry.GetNextAncestorBinary(), Snapshot, shadowInsertion)
+	if err != nil || newProcessNode && shadowInsertion {
```
Comment (safchain) [inline pkg/security/security_profile/activity_tree/activity_tree.go:287]: just : `if err != nil || shadowInsertion` ?

File: pkg/security/security_profile/activity_tree/activity_tree_secl.go [hunk 6]
```diff
@@ -6,56 +6,37 @@
...
 
+	"github.com/DataDog/datadog-agent/pkg/security/secl/compiler/eval"
 	"github.com/DataDog/datadog-agent/pkg/security/secl/model"
 	"github.com/DataDog/datadog-agent/pkg/security/utils"
 )
 
-// Profile holds the list of rules generated from an activity dump
-type Profile struct {
+// SeclEncoding holds the list of rules generated from an activity dump
+type SeclEncoding struct {
```
Comment (safchain) [inline pkg/security/security_profile/activity_tree/activity_tree_secl.go:20]: Wondering if shouldn't get rid of this ?

File: pkg/security/probe/probe_linux.go (hunk 1 above)
Comment (Gui774ume) [reply pkg/security/probe/probe_linux.go:324]: The goal was to better differentiate the dedicated anomaly detection event types generated in kernel space (this is a hardcoded list of events) from the user configurable list of events that can generate anomalies. The first one is also used to implement some fast paths such as in `LookupEventInProfiles`.

Last but not least, I think there is value in making this list dynamic and per profile (from the configurable default) in case we want to control the "load" of anomalies by selectively preventing a workload from generating anomalies for specific event types; or simply because the content of a profile do not allow for generating anomalies on some event types (we're missing this parameter in profiles right now and we should add it otherwise the transition between profiles with `chmod` support and without `chmod` support will be messy for example).

Does it make sense ?

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+)
+
+func (genType NodeGenerationType) String() string {
+	switch genType {
+	case Runtime:
+		return "runtime"
+	case Snapshot:
+		return "snapshot"
+	default:
+		return "unknown"
```
[the remark this answers is not in the document]
Comment (Gui774ume) [reply pkg/security/security_profile/activity_tree/activity_tree.go:58]: Thanks ! It's a mistake and I even forgot to propagate the `SecurityProfile` node generation type everywhere in the tree !

File: pkg/security/security_profile/activity_tree/activity_tree.go (hunk 2 above)
Comment (Gui774ume) [reply pkg/security/security_profile/activity_tree/activity_tree.go]: 👍🏻

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+		}
+	}()
+
+	// check if this event type is traced
+	if valid, err := at.isEventValid(event, shadowInsertion); !valid || err != nil {
+		return false, fmt.Errorf("invalid event: %s", err)
+	}
+
+	// find the node where the event should be inserted
+	entry, _ := event.FieldHandlers.ResolveProcessCacheEntry(event)
```
[the remark this answers is not in the document]
Comment (Gui774ume) [reply pkg/security/security_profile/activity_tree/activity_tree.go:216]: 👍🏻

File: pkg/security/security_profile/activity_tree/activity_tree.go (hunk 3 above)
Comment (Gui774ume) [reply pkg/security/security_profile/activity_tree/activity_tree.go:266]: ok removing it then 👍🏻

File: pkg/security/security_profile/activity_tree/activity_tree.go (hunk 5 above)
Comment (Gui774ume) [reply pkg/security/security_profile/activity_tree/activity_tree.go:294]: I think the `newProcessNode` check is important: unless I'm missing something, we want to check if we're modifying something in the graph. This includes making sure that the entire process tree is in the profile. We can return early as soon as (and only if) we see something new.

File: pkg/security/security_profile/profile/manager.go [hunk 9]
```diff
@@ -511,62 +531,73 @@ func (m *SecurityProfileManager) unlinkProfile(profile *SecurityProfile, workloa
...
-	if evtType == model.SyscallsEventType || // syscall matching for anomaly detection is already done kernel side
-		evtType == model.FileOpenEventType || evtType == model.BindEventType || // disabled for now
-		evtType == model.ForkEventType || evtType == model.ExitEventType { // no interest in fork/exit events
-		return
-	}
-
-	if event.Error != nil {
-		m.eventFilteringAbsent[evtType].Inc()
+func (m *SecurityProfileManager) LookupEventInProfiles(event *model.Event) {
+	// shortcut for dedicated anomaly detection events
```
Comment (spikat) [inline pkg/security/security_profile/profile/manager.go:535]: I think we should also have some shortcuts here to check first if the event is a valid one.
You move both checks (event types and errors) on profile:isEventValid(), which is only call later on when trying to insert or checking if the event is present on a profile.
This could avoid unneeded calls to resolvers, workloadselector creation, profiles lookup etc

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+	entry, isResolved := event.FieldHandlers.ResolveProcessCacheEntry(event)
+	if !isResolved {
+		return false, fmt.Errorf("process cache entry not found")
+	}
+	if !entry.HasCompleteLineage() && !shadowInsertion { // check that the process context lineage is complete, otherwise drop it
+		at.Stats.droppedCount[event.GetEventType()][brokenLineageReason].Inc()
+		return false, fmt.Errorf("incomplete lineage")
+	}
+
+	node, newProcessNode, err := at.CreateProcessNode(entry, generationType, shadowInsertion)
```
Comment (spikat) [inline pkg/security/security_profile/activity_tree/activity_tree.go:225]: IMO we should only CreateProcessNode if !shadowInsertion

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+	if newProcessNode && shadowInsertion {
+		return true, nil
+	}
+	if err != nil {
```
Comment (spikat) [inline pkg/security/security_profile/activity_tree/activity_tree.go:229]: nit: we may want to check err before checking newProcessNode && shadowInsertion, right?

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+	if err != nil || newProcessNode && shadowInsertion {
+		return parentNode, newProcessNode, err
+	}
+
+	// if parentNode is nil, the parent of the current node is out of tree (either because the parent is null, or it
+	// doesn't match the dump tags).
+	if parentNode == nil {
+
+		// since the parent of the current entry wasn't inserted, we need to know if the current entry needs to be inserted.
+		if !at.validator.MatchesSelector(entry) {
```
Comment (spikat) [inline pkg/security/security_profile/activity_tree/activity_tree.go:299]: I wonder if we may want to do this check sooner (to avoid searching for a parent if the selector doesnt match), what did you think?

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 13]
```diff
@@ -0,0 +1,381 @@
...
+			return nil, false, nil
+		}
+
+		// go through the root nodes and check if one of them matches the input ProcessCacheEntry:
```
Comment (spikat) [inline pkg/security/security_profile/activity_tree/activity_tree.go:303]: I wonder if we should make this search before searching for parents.

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 14]
```diff
@@ -0,0 +1,381 @@
...
+		// we're about to add a root process node, make sure this root node passes the root node sanitizer
+		if !at.isValidRootNode(&entry.ProcessContext) {
+			if !shadowInsertion {
+				at.Stats.droppedCount[model.ExecEventType][invalidRootNodeReason].Inc()
+			}
+			return nil, false, fmt.Errorf("invalid root node")
+		}
+
+		// if it doesn't, create a new ProcessActivityNode for the input ProcessCacheEntry
+		node = NewProcessNode(entry, generationType)
```
Comment (spikat) [inline pkg/security/security_profile/activity_tree/activity_tree.go:319]: same remark as before, IMO we should not call NewProcessNode if !shadowInsertion

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+		// to add the current entry no matter if it matches the selector or not. Go through the root children of the
+		// parent node and check if one of them matches the input ProcessCacheEntry.
+		for _, child := range parentNode.Children {
+			if child.Matches(&entry.Process, at.differentiateArgs) {
+				return child, false, nil
+			}
+		}
+
+		// if none of them matched, create a new ProcessActivityNode for the input processCacheEntry
+		node = NewProcessNode(entry, generationType)
```
Comment (spikat) [inline pkg/security/security_profile/activity_tree/activity_tree.go:338]: Same remark here

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 16]
```diff
@@ -0,0 +1,381 @@
...
+			at.Stats.ProcessNodes++
+		}
+
+	} else {
+
+		// if parentNode wasn't nil, then (at least) the parent is part of the activity dump. This means that we need
+		// to add the current entry no matter if it matches the selector or not. Go through the root children of the
+		// parent node and check if one of them matches the input ProcessCacheEntry.
+		for _, child := range parentNode.Children {
+			if child.Matches(&entry.Process, at.differentiateArgs) {
```
Comment (spikat) [inline pkg/security/security_profile/activity_tree/activity_tree.go:332]: For both calls to Matches() (also on line 305):
Here the behavior changes from what we had before: we use to set this flags from config file for activity dumps, and force it to false for security profiles.
Making this generic would lead into some possible inconsistency: if some profiles are generated with DA yes (or no), then merged into a profile and send it back to the agent (or having an agent config change at some point), the loading agent could have DA no (or yes); and we will have a difference between the way we constructed the profile, and the way we handle it afterward. (sorry if I'm not clear enough here :'D)
tl;dr: I would rather prefer to keep the loose approach for profiles, by keeping this flags to false.
For the activity dump part, we could either keep it as it (choosing one of the two modes from the config) .. or, I wonder if we just can .. force it to false too. what did you think?

**Review by spikat — COMMENTED** [review]
I've added some comments, but nice PR overall! :)

File: pkg/security/security_profile/profile/manager.go [hunk 17]
```diff
@@ -511,62 +531,73 @@ func (m *SecurityProfileManager) unlinkProfile(profile *SecurityProfile, workloa
...
+	if IsAnomalyDetectionEvent(event.GetEventType()) {
+		event.AddToFlags(model.EventFlagsSecurityProfileInProfile)
 		return
 	}
 
-	event.FieldHandlers.ResolveContainerID(event, &event.ContainerContext)
```
[the remark this answers is not in the document]
Comment (Gui774ume) [reply pkg/security/security_profile/profile/manager.go:527]: Unless I'm missing something, I don't think we need it here 🤔 To be safe I added it in the `handleAnomalyDetection` function in case we're looking up a profile by container ID.

File: pkg/security/security_profile/activity_tree/activity_tree.go (hunk 10 above)
Comment (Gui774ume) [reply pkg/security/security_profile/activity_tree/activity_tree.go:225]: Actually this call is needed because we use it to go up the process tree and check the process lineage ... but I'll think again about not calling `New*Node` ... as you suggested later we might not need it.

File: pkg/security/security_profile/activity_tree/activity_tree.go (hunk 11 above)
Comment (Gui774ume) [reply pkg/security/security_profile/activity_tree/activity_tree.go:229]: 👍🏻

File: pkg/security/security_profile/activity_tree/activity_tree.go (hunk 12 above)
Comment (Gui774ume) [reply pkg/security/security_profile/activity_tree/activity_tree.go:299]: So the good news here is that the cookie to process shortcut ensures that we only go up the process tree if needed, including the parent check you're talking about 🎉 
In other words, either the process as already been seen before and therefore we'll never reach the part about the parent process, or it hasn't been seen before, and then the optimal algorithm is:
- to go look for the first process that has been seen before (so that you know where to start in the tree)
- or that is at the root of the process tree and hasn't been seen before

Let's chat if it doesn't make sense :)

File: pkg/security/security_profile/activity_tree/activity_tree.go (hunk 13 above)
Comment (Gui774ume) [reply pkg/security/security_profile/activity_tree/activity_tree.go:303]: Happy to chat about it, but if the process in question is a root node, then this is almost what the algorithm will do first. More importantly I don't really think there is any statistical reason why a process generating events in a profile is most likely to be a root process, so I would actually stick to the generic solution as it would prevent going over the same nodes in the tree multiple times.

File: pkg/security/security_profile/profile/manager.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (spikat) [reply pkg/security/security_profile/profile/manager.go:527]: Yeah, I wasn't sure, I just saw that we had it before and that was removed. It should be fine, I resolve :)

File: pkg/security/security_profile/profile/manager.go (hunk 9 above)
Comment (spikat) [reply pkg/security/security_profile/profile/manager.go:535]: Yep it could be error prone.. Maybe we could call isEventValid() directly from both dump and profile sides?

File: pkg/security/security_profile/activity_tree/activity_tree.go (hunk 13 above)
Comment (spikat) [reply pkg/security/security_profile/activity_tree/activity_tree.go:303]: Yep, here again I forgot to think about it with the cookies in mind. I was afraid to maybe do this check twice on the call path (for the node and its parent). But with cookies it should not happen, you're right!

File: pkg/security/security_profile/profile/manager.go [hunk 18]
```diff
@@ -511,62 +531,73 @@ func (m *SecurityProfileManager) unlinkProfile(profile *SecurityProfile, workloa
...
-		event.AddToFlags(model.EventFlagsSecurityProfileInProfile)
+	// check if the event should be injected in the profile automatically
+	if profile.autolearnEnabled {
+		if profile.lastAnomalyNano == 0 {
+			profile.lastAnomalyNano = event.TimestampRaw
+		}
+		if time.Duration(event.TimestampRaw-profile.lastAnomalyNano) >= m.config.RuntimeSecurity.AnomalyDetectionMinimumStableDelayBeforeAnomalyDetection {
+			profile.autolearnEnabled = false
+		} else {
+			newEntry, err := profile.ActivityTree.Insert(event, activity_tree.SecurityProfile)
```
[the remark this answers is not in the document]
Comment (Gui774ume) [reply pkg/security/security_profile/profile/manager.go]: Done ! Thanks 👍🏻

File: pkg/security/security_profile/profile/manager.go [hunk 19]
```diff
@@ -511,62 +537,100 @@ func (m *SecurityProfileManager) unlinkProfile(profile *SecurityProfile, workloa
...
+
+	// have we reached the unstable time limit ?
+	if time.Duration(event.TimestampRaw-profile.loadedNano) >= m.config.RuntimeSecurity.AnomalyDetectionUnstableProfileTimeThreshold {
+		m.eventFiltering[event.GetEventType()][UnstableProfile].Inc()
+		return false, fmt.Errorf("unstable profile: time limit reached")
+	}
+
+	// check if the unstable size limit was reached
+	if profile.ActivityTree.Stats.ApproximateSize() >= m.config.RuntimeSecurity.AnomalyDetectionUnstableProfileSizeThreshold {
+		m.eventFiltering[event.GetEventType()][UnstableProfile].Inc()
```
Comment (spikat) [inline pkg/security/security_profile/profile/manager.go:613]: should we add `profile.autolearnEnabled = false` when we reach the safe guards ? (here and on the previous if)

**Review by spikat — APPROVED** [review]
LGTM

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 20]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+// isValidRootNode evaluates if the provided process entry is allowed to become a root node of an Activity Dump
+func (at *ActivityTree) isValidRootNode(entry *model.ProcessContext) bool {
+	// TODO: evaluate if the same issue affects other container runtimes
+	return !(strings.HasPrefix(entry.FileEvent.BasenameStr, "runc") || strings.HasPrefix(entry.FileEvent.BasenameStr, "containerd-shim"))
+}
+
+// isEventValid evaluates if the provided event is valid
+func (at *ActivityTree) isEventValid(event *model.Event, shadowInsertion bool) (bool, error) {
```
Comment (lebauce) [inline pkg/security/security_profile/activity_tree/activity_tree.go:153]: s/shadowInsertion/dryRun/ ?

File: pkg/security/security_profile/activity_tree/activity_tree_secl.go [hunk 21]
```diff
@@ -6,56 +6,37 @@
...
-	"text/template"
 
+	"github.com/DataDog/datadog-agent/pkg/security/secl/compiler/eval"
 	"github.com/DataDog/datadog-agent/pkg/security/secl/model"
 	"github.com/DataDog/datadog-agent/pkg/security/utils"
 )
 
-// Profile holds the list of rules generated from an activity dump
-type Profile struct {
+// SeclEncoding holds the list of rules generated from an activity dump
```
Comment (lebauce) [inline pkg/security/security_profile/activity_tree/activity_tree_secl.go:19]: nit: SECLEncoding

**Review by safchain — APPROVED** [review]

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 22]
```diff
@@ -0,0 +1,382 @@
...
+	return at
+}
+
+// ComputeSyscallsList computes the top level list of syscalls
+func (at *ActivityTree) ComputeSyscallsList() []uint32 {
+	output := make([]uint32, 0, len(at.SyscallsMask))
+	for key := range at.SyscallsMask {
+		output = append(output, uint32(key))
+	}
+	return output
```
Comment (paulcacheux) [inline pkg/security/security_profile/activity_tree/activity_tree.go:115]: Should we sort the output ? AFAIK iterating over a map (SyscallsMask) is in random order

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 23]
```diff
@@ -0,0 +1,382 @@
...
+	for _, node := range at.ProcessNodes {
+		if node.Process.FileEvent.BasenameStr == basename {
+			res = append(res, node)
+		}
+	}
+	return res
+}
+
+// Snapshot uses procfs to snapshot the nodes of the tree
+func (at *ActivityTree) Snapshot(newEvent func() *model.Event) interface{} {
```
Comment (paulcacheux) [inline pkg/security/security_profile/activity_tree/activity_tree.go:368]: shouldn't this return `error` ?

**Review by paulcacheux — COMMENTED** [review]
Few nits, but otherwise LGTM

File: pkg/security/security_profile/activity_tree/activity_tree.go (hunk 22 above)
Comment (safchain) [reply pkg/security/security_profile/activity_tree/activity_tree.go:115]: I'm not sure the order matter but I will sort it, it will be easier if we want to compare

**Review by paulcacheux — APPROVED** [review]
